cmd/agent/app/throttling/credit_store.go

```diff
@@ -0,0 +1,208 @@
+	balance          float64
+	maxBalance       float64
+	lastTick         time.Time
+	timeNow          func() time.Time // For testing
+}
+
+func (t *tokenBucket) Drain(maxDrain float64) float64 {
+	now := t.timeNow()
+	interval := now.Sub(t.lastTick)
+	t.lastTick = now
+	diff := interval.Seconds() * t.creditsPerSecond
+	t.balance = math.Min(t.balance+diff, t.maxBalance)
+	result := math.Min(t.balance, maxDrain)
+	t.balance -= result
+	return result
+}
+
+type client struct {
+	id                  string
+	perOperationBalance map[string]float64
+	defaultBalance      float64
+	updateTime          time.Time
+}
+
+func (c *client) Spend(operationName string, credits float64) bool {
+	balance, ok := c.perOperationBalance[operationName]
+	if !ok {
+		balance = c.defaultBalance
+	}
+
+	if credits > balance {
+		return false
+	}
+
+	balance -= credits
+	if ok {
+		c.perOperationBalance[operationName] = balance
+	} else {
+		c.defaultBalance = balance
+	}
+	return true
+}
+
+type CreditAccruerOptions struct {
+	MaxOperations    int
+	CreditsPerSecond float64
+	MaxBalance       float64
+	ClientMaxBalance float64
+}
+
+type creditAccruer struct {
+	options                 CreditAccruerOptions
+	serviceName             string
+	perOperationRateLimiter map[string]*tokenBucket
+	defaultRateLimiter      *tokenBucket
+	updateTime              time.Time
+	timeNow                 func() time.Time // For testing
+}
+
+func (ca *creditAccruer) Withdraw(operationName string, balance float64) float64 {
+	if balance >= ca.options.ClientMaxBalance {
+		return 0
+	}
+	rateLimiter, ok := ca.perOperationRateLimiter[operationName]
+	if !ok {
+		if len(ca.perOperationRateLimiter) < ca.options.MaxOperations {
+			rateLimiter = &tokenBucket{
+				creditsPerSecond: ca.options.CreditsPerSecond,
+				balance:          ca.options.MaxBalance,
+				maxBalance:       ca.options.MaxBalance,
+				timeNow:          ca.timeNow,
+				lastTick:         ca.timeNow(),
+			}
+			ca.perOperationRateLimiter[operationName] = rateLimiter
+		} else {
+			rateLimiter = ca.defaultRateLimiter
+		}
+	}
+
+	credits := rateLimiter.Drain(ca.options.ClientMaxBalance - balance)
+	return credits
+}
+
+type creditStore struct {
+	creditAccruers map[string]*creditAccruer
+	clients        map[string]*client
+	defaultOptions CreditAccruerOptions
+	lifetime       time.Duration
+	timeNow        func() time.Time // For testing
+}
+
+func NewCreditStore(defaultOptions CreditAccruerOptions, lifetime time.Duration) *creditStore {
+	s := &creditStore{
+		creditAccruers: map[string]*creditAccruer{},
+		clients:        map[string]*client{},
+		defaultOptions: defaultOptions,
+		lifetime:       lifetime,
+	}
+	s.timeNow = time.Now
+	return s
+}
+
+func (s *creditStore) Withdraw(clientID string, serviceName string, operationName string) float64 {
+	c := s.findOrCreateClient(clientID)
+	ca := s.findOrCreateCreditAccruer(serviceName)
+
+	// To keep client consistent with creditAccruer, we need to see if creditAccruer
```

Comment:
I don't see a big reason to keep these two consistent. I think it's fine for the client to track an operation without it having an entry in the creditAccruer. (ie most clients before we roll this out everywhere). This can be simplified to:

```
balance = c.getBalance() <- create a per operation balance if needed
ca.Widthdraw(operationName, clientMaxBalance) <- if operation isn't being tracked, create a new one or use default
```

Reply:
I guess the TTL makes this a non-issue, but my concern was the growth of keys for no apparent reason. Also, it seems sensible for me for clients to have the correct idea of what is going on in the rate limiter, not to expect per operation for every new operation. I can change it, but I'll try to see if I can simplify the logic and leave the behavior the same.